src/main/java/org/eclipse/uprotocol/streamer/UStreamer.java

```diff
@@ -0,0 +1,128 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.eclipse.uprotocol.streamer;
+
+import org.eclipse.uprotocol.v1.UCode;
+import org.eclipse.uprotocol.v1.UStatus;
+import org.eclipse.uprotocol.v1.UUri;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/** 
+ * Example implementation of "Pluggable Streamer" written in Java.
+ * 
+ * The streamer provides APIs to simply add remove forwarding rules. A rule is defined as a 
+ * combination of input and output routes. The input route is the route to listen to and the output route is the route to forward the message to.
+ * The forwarding works by registering a listener on the input route and forwarding the message to the output route.
+ */
+public class UStreamer {
+
+    // List of listeners (routes) that the streamer listens to
+    private List<TransportListener> listeners;
+
+
+    /**
+     * Constructor
+     */
+    public UStreamer() {
+        listeners = new ArrayList<>();
+    }
+
+
+    /**
+     * Add a forwarding rule to the streamer. 
+     * The input route is the route to listen to and the output route is the route to forward the message to.
+     * The forwarding works by registering a listener on the input route and forwarding the message to the output route.
+     * 
+     * NOTE: Default routing rule (where UAuthority.name is "*") *MUST* be added at the end of the list to avoid overriding other rules.
+     * 
+     * @param in input {@code Route} that the streamer listens to
+     * @param out output {@code Route} that the streamer forwards the message to
+     * @return {@code UStatus} with UCode.OK if the forwarding rule was added successfully,
+     *        {@code UStatus} with UCode.ALREADY_EXISTS if the forwarding rule already exists,
+     *       {@code UStatus} with UCode.INVALID_ARGUMENT if the input and output routes are the same
+     */
+    public UStatus addForwardingRule(Route in, Route out) {
+        Objects.requireNonNull(in, "input cannot be null.");
+        Objects.requireNonNull(out, "output cannot be null.");
+        
+        // Cannot route to itself
+        if (in.equals(out)) {
+            return UStatus.newBuilder().setCode(UCode.INVALID_ARGUMENT).build();
+        }
+
+        // check if the rule already exists in the list
+        if (listeners.stream().anyMatch(p -> p.getInputRoute().equals(in) && p.getOutputRoute().equals(out))) {
+            return UStatus.newBuilder().setCode(UCode.ALREADY_EXISTS).build();
+        }
+
+        TransportListener listener = new TransportListener(in, out);
+        UUri out_uri = UUri.newBuilder().setAuthority(out.getAuthority()).build();
+
+        UStatus result = in.getTransport().registerListener(out_uri, listener);
```

Comment:
I am recalling @sophokles73's [feedback from about a week ago](https://github.com/eclipse-uprotocol/up-spec/issues/73#issuecomment-2004462119) as I was implementing `up-streamer-rust`.

It seems to me this would nab you the Request, Response, and Notification messages headed to `out.getAuthority()`, but not Publish messages _originating from_ `in.getAuthority()`.

I think we may need to modify:
```java
UStatus result = in.getTransport().registerListener(out_uri, request_response_notification_listener);
```

and add:
```java
UUri in_uri = UUri.newBuilder().setAuthority(in.getAuthority()).build();
UStatus result = in.getTransport().registerListener(in_uri, publish_listener);
``` 
and make some corresponding changes to `TransportListener` to allow us to configure it for dropping Publish messages or all non-Publish messages.

Still tooling around with this over on `up-streamer-rust`, but wanted to make this comment before I forget :slightly_smiling_face: